### ISSUE TITLE
Could we add a  Cluster Distributor to dask_feature_extraction_on_chunk  also?

### ISSUE THREAD

**Issue by demontamer** [blue-yonder/tsfresh#653]
I added the parameter in #648

_Originally posted by @nils-braun in https://github.com/blue-yonder/tsfresh/issues/646#issuecomment-603970797_

Hi @nils-braun , I have had difficulty extracting features from dask dataframes unitl now i noticed recently you added a new dask_feature_extraction function, which is super amazing, now can you add a cluster distributor to this as well? 

Cheers

**Comment by demontamer** [comment]
To provide more background on the issues I run into with dask and tsfresh...
1. I set up a cluster on Amazon EMR.
2. I tried the following copied and paste from Pycon 2017 presentation. So..
...`melted = pd.melt(df)`
`melted.rename(columns={"variable": "kind", "value": "price"}, inplace=True)`
`melted["time"] = list(df.index)*len(df.columns)`
`melted["id"] = 1`
`from tsfresh.utilities.dataframe_functions import roll_time_series`
`use_data_of = 100`
`rolled = roll_time_series(melted.copy(), column_id="id", column_sort="time",
 column_kind="kind",
 rolling_direction=1,
 max_timeshift=use_data_of - 1).reset_index(drop=True)`
`rolled = rolled.groupby("id")\
 .filter(lambda x: len(set(x["time"])) == use_data_of)\
 .reset_index(drop=True)`

Now i create my distributor then try to extract features
`Distributor = ClusterDaskDistributor('tcp://172..')`
`X = extract_features(rolled, column_id="id", column_sort="time",
 column_kind="kind", column_value="price",
 default_fc_parameters=EfficientFCParameters(), distributor=Distributor)`

This causes and error similiar to:
`MemoryError: Unable to allocate array with shape (3, 156082680) and data type float64`
After some research found [this answer](https://docs.dask.org/en/latest/best-practices.html#load-data-with-dask) from @mrocklin. Essential better to load data as dask dataframe. Is this correct?

As you may know using dask taskes some experience so had lots of errors to do with column assignments and groupings. etc. Eventually I tried converting rolled above after the grouping to a dask array like:
`rolled = dd.from_pandas(rolled, npartitions=2)` 
Now I ran into;
`Traceback (most recent call last):`
  `File "app/trading/mlbots/test/tsf_dask.py", line 120, in <module>`
    `get_data('s3://test_cdr2usd.csv')`
`  File "app/test/tsf_dask.py", line 98, in get_data`
   ` default_fc_parameters=EfficientFCParameters(),distributor=Distributor) `
  `File "/usr/local/lib/python3.6/site-packages/tsfresh/feature_extraction/extraction.py", line 135, in extract_features`
    `column_value=column_value)`
  `File "/usr/local/lib/python3.6/site-packages/tsfresh/utilities/dataframe_functions.py", line 295, in _normalize_input_to_internal_representation`
  `  if timeseries_container[column_id].isnull().any():`
  `File "/usr/local/lib/python3.6/site-packages/dask/dataframe/core.py", line 196, in __bool__
    "a conditional statement.".format(self)`
`TypeError: Trying to convert dd.Scalar<series-..., dtype=bool> to a boolean value. Because Dask objects are lazily evaluated, they cannot be converted to a boolean value or used in boolean conditions like if statements. Try calling .compute() to force computation prior to converting to a boolean value or using in a conditional statement.`


Also tried creating a dask array directly from csv file plus working round above mentioned issues with using same grouping above.  But kept got similar error a above with

`File "/usr/local/lib/python3.6/site-packages/tsfresh/feature_extraction/extraction.py", line 135, in extract_features
    column_value=column_value)`

Finally. saw the convenience  fundctions recently for dask feature extraction on chunk.

1.  However how do we use this on a cluster? 
2. Extract features based on maximum time shift?

**Comment by nils-braun** [comment]
Hi!
So first of all, thanks for trying out the new dask binding before it is even officially announced :-) There will be more information following up once I have some time to write something....
Second: your bug report was very hard to read - I just added some formatting to it. Please make sure to have code formatted as code - and if you see that the github issue does not look good please try to edit it afterwards. Thanks!

Now to your issue:
* yes, a dataframe with a dimension of 156082680 will probably not fit into your memory (and also not be possible to work with) ;-) So this is a very good reason to use dask
* tsfresh is built around pandas dataframes. You can not "just" input dask dataframes into tsfreshs ''extract_features'` function - that will not work. So you either need to to convert them to pandas dataframes (but this will bring us to the memory problem again..) or:
* You are using the new (still unannounced) function `dask_feature_extraction`- In this case, you start with a dask dataframe, then apply the function (see the documentation here: https://tsfresh.readthedocs.io/en/master/api/tsfresh.convenience.html#module-tsfresh.convenience.bindings) and still have a dask dataframe afterwards. You can distribute this to your cluster as you would with a normal dask dataframe. Please have a look into the dask documentation, which is super well written :-) https://docs.dask.org/en/latest/dataframe.html

So in short:
1. start with a dask dataframe, e.g. reading it from s3
2. then do the dataframe manipulation etc. with dask
3. if you have reached the input form for tsfresh (melting, rolling, whatever you need with dask), call the `dask_feature_extraction` as written in the documentation.
4. Then do whatever you want with the result (which is again a dask dataframe) - e.g. store it somewhere.

In the end, call the calculation and distribute it with the normal dask possibilities.
Happy coding!

(If you have a more specific issue after you tried out starting with dask, please post it here and re-open!)

**Comment by demontamer** [comment]
Hi Nils, thanks for the reply. Sorry if my formating skills are not up to scratch. I did as you said and started with dask . You can see this on the 6th block of code from above and the resulting error in the following block. Also used a dask array then applied dask_feature_extraction which worked. So really just trying to find out how to distribute dask_feature_extraction to a cluster?

**Comment by nils-braun** [comment]
Hi @demontamer!
Ok I see, let me try to help you on the dask part (although I guess I am also not really an expert here):
1. You need a dask cluster with workers (or at least a local worker for testing). There exist multiple ways on how you can create one and it mostly depends on which environement you are using. A general description is here: https://distributed.dask.org/en/latest/quickstart.html, but there is a lot more ways described in here: https://docs.dask.org/en/latest/setup.html
2. When you now start your python script/ipython shell/jupyter notebook you need to call 
```python
from dask.distributed import Client
client = Client(...)  
```
  as described here: https://docs.dask.org/en/latest/setup.html. This will make all subsequent dask calculations running in cluster mode. The parameters to the `Client` call depend on the setup you have chosen.
3. Now create your dask dataframe just as you have done above, but do not put it into `extract_features` but use the `dask_feature_extraction` (I guess you said you have done so, but her is the docu again: https://tsfresh.readthedocs.io/en/master/api/tsfresh.convenience.html#module-tsfresh.convenience.bindings=
4. Trigger some action in the resulting dask data frame, e.g. write it out. This will then trigger the full calculation, which will run on the cluster (as you have imported the client above).

> And to report that just adding a ClusterDaskDistributor parameter to extract_features doesn't work for an actual cluster.
This part I did not quite understand. Please note, that this is unrelated to the problems you have described above - in my description I am talking how you can input a dask dataframe into tsfresh. The `ClusterDaskDistributor` is used when you want to parallelize the internal calculation of tsfresh with dask (which is not the problem you probably need to solve).

**Comment by demontamer** [comment]
>This part I did not quite understand. Please note, that this is unrelated to the problems you have described above - in my description I am talking how you can input a dask dataframe into tsfresh. The ClusterDaskDistributor is used when you want to parallelize the internal calculation of tsfresh with dask (which is not the problem you probably need to solve).

Hi, this is exactly the problem that 'am trying to solve. Using ClusterDaskDistributor to parallelize the feature extraction calculations. I set up an EMR cluster on AWS using https://yarn.dask.org/en/latest/aws-emr.html and got the address of the scheduler as parameter for your `ClusterDaskDistributor` class. Now the piece of code below works for `Distributor = LocalDaskDistributor(n_workers=8)` but I ran into the memory error described above when I use the EMR cluster. See:

```
Distributor = ClusterDaskDistributor(get_scheduler())
X = extract_features(rolled, column_id="id", column_sort="time",
 column_kind="kind", column_value="price",
 default_fc_parameters=EfficientFCParameters(), distributor=Distributor)
```
So when I investigated, it looked like the distributed cluster wants a dask array not pandas. Which is why I thought I needed a dask array? Can you look into it? You only need to setup an EMR cluster as in the link above to reproduce the error, I think.

```
def get_scheduler():
    cluster = YarnCluster()
    cluster.scale(2)
    return str(cluster.scheduler_address)
```
The real question is why does with it work fine with a LocalDaskDistributor but not with ClusterDaskDistributor? Thanks

**Comment by nils-braun** [comment]
Hi @demontamer!

I think there is a lot going on in this issue, so let me try to clarify a few things :-)

(first of all thanks for formating the code in your answer, looks very good!)

There are two possibilities implemented in `tsfresh`, how you can use `dask` and they are extremely different. Before we go into the issues, let my briefly summarize that:
* when you are using a distributor (either `LocalDaskDistributor` or `ClusterDaskDistributor`, does not matter) you are parallelizing the calculation of the features, but you are still (important!) using pandas dataframes all the time. This means all your data needs to fit into memory of your driver node at any given time. This is the number one drawback, but the benefit is that you can still use your "usual" pandas operations. 
* when you are using the dask bindings (`dask_feature_extraction`) you start with a dask dataframe (not pandas), which means you have distributed data (not only calculation). Dask will take care to only have the memory allocated on the worker nodes for the amount of data that you actually need. You start and end with a dask dataframe - there is no pandas involved (at least from the outside). This possibility gives you much more flexibility and scalability, but comes with additional burdens of using a distributed framework (e.g. you need to read in and write the data with dask, not with pandas).

Let me repeat again: distributor = parallel computation, still use pandas (all data need to fit into memory). dask bindings = start and end with a dask dataframe, no pandas involved.
Sorry that the documentation is still sparse on the dask binding - I am still working on this part :-)

Now to your issue: the error message you posted happens in the `generate_data_chunk_format` function. This one is called *before* there is actually any distributor involved (before we do any parallelization or distribution) - so there should be no reason why there is a difference between the two distributors. Only reason I can think if is that creating a cluster distributor costs more memory and less memory is available for the rest of the calculation.
Three questions/suggestions.
* make sure you are calling this from a script and not from an interactive console to make sure really just the data that needs to be in memory are allocated. For example if you want to test out both the local and the cluster distributor, do not do this in the same script or interactive shell and do not do this simultanously. Make sure to clean up after the first test (e.g. by closing the python shell or exiting the program) before trying the next one. Just to be sure :-) 
* could you just try to instantiate a cluster distributor without doing the calculation and check how the memory consumption goes up? It is strange that there is such a large difference.
* if it really turns out that memory is your problem (even though you said it works for the local distributor and I guess you tested the same amount of data in both possibilities, otherwise it is clear why there is a difference :-)), please go with the dask bindings (not the distributor) as I tried to make clear in my post above and in this post.

Hope this helps you!

**Comment by demontamer** [comment]
Thanks for clearing things up. If it occurs before parallelization, then I'll investigate further or see if adding more memory or even a swapfile works. In the meantime I'll close this issue or reopen if neccessary. Cheers.

**Comment by demontamer** [comment]
Ok. So before I used a tiny bit of data just 2000 rows. And had the issues above. Now I increased my cluster size to 32cores and 2GB but still had problems similar to this https://stackoverflow.com/questions/47776936/why-is-a-computation-much-slower-within-a-dask-distributed-worker

FYI: Setting chained assignment to none using `pd.options.mode.chained_assignment = None` didn't work for me.

So, I now try to return to using the local cluster `Distributor = LocalDaskDistributor(n_workers=32)` but this time increasing chunk of data to 10000 I have over 100000 rows. `data = data[50000:60000]
`
But alas the original error came back even with a Local cluster
```
/home/hadoop/miniconda/lib/python3.7/site-packages/tsfresh/utilities/dataframe_functions.py:460: UserWarning: Your time stamps are not uniformly sampled, which makes rolling nonsensical in some domains.
  warnings.warn("Your time stamps are not uniformly sampled, which makes rolling "
Traceback (most recent call last):
  File "tsf_dask.py", line 123, in <module>
    get_data('s3://test_cdr2usd.csv')
  File "tsf_dask.py", line 77, in get_data
    column_kind="kind",rolling_direction=1,max_timeshift=use_data_of - 1).reset_index(drop=True)
  File "/home/hadoop/miniconda/lib/python3.7/site-packages/tsfresh/utilities/dataframe_functions.py", line 491, in roll_time_series
    df_shift = pd.concat([roll_out_time_series(time_shift) for time_shift in range_of_shifts], ignore_index=True)
  File "/home/hadoop/miniconda/lib/python3.7/site-packages/pandas/core/reshape/concat.py", line 284, in concat
    return op.get_result()
  File "/home/hadoop/miniconda/lib/python3.7/site-packages/pandas/core/reshape/concat.py", line 497, in get_result
    mgrs_indexers, self.new_axes, concat_axis=self.axis, copy=self.copy
  File "/home/hadoop/miniconda/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 2018, in concatenate_block_managers
    [ju.block for ju in join_units], placement=placement
  File "/home/hadoop/miniconda/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 359, in concat_same_type
    [blk.values for blk in to_concat], axis=self.ndim - 1
  File "<__array_function__ internals>", line 6, in concatenate
MemoryError: Unable to allocate 28.9 GiB for an array with shape (1, 3882548250) and data type object
```
Now trying to see if this works `$ echo 0 | sudo tee /proc/sys/vm/overcommit_memory `

**Comment by nils-braun** [comment]
Aha, nie I understand. You are doing a time series role before the actual feature extraction, correct?
Well, time series rolling needs a lot of memory and is currently only implemented for pandas dataframes (so dask will not help you here).
I see two possibilities:
* you implement the rolling for dask and use the distributed system of dask to do the computation, which will make sure not too much memory is used at a given time or
* You do the reading of data, rolling, feature extraction und storing of the results somehow one-time-series-at-a-time. This means you always only have a small amount of data in memory, do all your pre-processing and feature extraction and then store it to disk releasing the allocated memory. You could use a task scheduler for orchestrating this.

For your information: I have finished the blog post on this topic. Especially part II (https://nils-braun.github.io/tsfresh-on-cluster-2/) might be interesting for you. In the end I also discuss how to do this step-by-step procedure (without rolling, but you should manage to include this by yourself I guess!)
